server/plugin/api.go

```diff
@@ -466,95 +481,199 @@ func (p *Plugin) handleDriveWatchNotifications(c *Context, w http.ResponseWriter
-		w.WriteHeader(http.StatusInternalServerError)
-		return
-	}
+	for _, change := range changes {
+		if change.File == nil {
+			continue
+		}
 
-	if activityRes == nil || activityRes.Activities == nil || len(activityRes.Activities) == 0 {
-		p.API.LogInfo("No activities found")
-		w.WriteHeader(http.StatusOK)
-		return
-	}
+		modifiedTime, _ := time.Parse(time.RFC3339, change.File.ModifiedTime)
+		lastChangeTime, _ := time.Parse(time.RFC3339, change.Time)
+		viewedByMeTime, _ := time.Parse(time.RFC3339, change.File.ViewedByMeTime)
 
-	activity := activityRes.Activities[0]
-	if activity.PrimaryActionDetail.Comment != nil {
-		p.handleCommentNotifications(lastChange.FileId, userID, activity, authToken)
-	}
-	if activity.PrimaryActionDetail.PermissionChange != nil {
-		p.handleFileSharedNotification(lastChange.FileId, userID, authToken)
+		// Check if the user has already opened the file after the last change.
+		if lastChangeTime.Sub(modifiedTime) > lastChangeTime.Sub(viewedByMeTime) {
+			continue
+		}
+
+		driveActivityQuery := &driveactivity.QueryDriveActivityRequest{
+			ItemName: fmt.Sprintf("items/%s", change.FileId),
+		}
+
+		lastActivityTime, err := p.KVStore.GetLastActivityForFile(userID, change.FileId)
+		if err != nil {
+			p.API.LogDebug("Failed to fetch last activity for file", "err", err, "fileID", change.FileId, "userID", userID)
+			continue
+		}
+
+		// If we have a last activity timestamp for this file we can use it to filter the activities.
+		if lastActivityTime != "" {
+			driveActivityQuery.Filter = "time > \"" + lastActivityTime + "\""
+		} else {
+			// PageSize documentation: https://developers.google.com/drive/activity/v2/reference/rest/v2/activity/query#QueryDriveActivityRequest.
+			// TLDR: PageSize does not return the exact number of activities that you specify. LastActivity is not set so lets just get the latest activity.
+			driveActivityQuery.PageSize = 1
+		}
+
+		var activities []*driveactivity.DriveActivity
+		for {
+			var activityRes *driveactivity.QueryDriveActivityResponse
+			activityRes, err = activitySrv.Activity.Query(driveActivityQuery).Do()
+			if err != nil {
+				p.API.LogError("Failed to fetch google drive activity", "err", err, "fileID", change.FileId, "userID", userID)
+				continue
+			}
+			activities = append(activities, activityRes.Activities...)
+			// NextPageToken is set when there are more than 1 page of activities for a file. We don't want the next page token if we are only fetching the latest activity.
+			if activityRes.NextPageToken != "" && driveActivityQuery.PageSize != 1 {
+				driveActivityQuery.PageToken = activityRes.NextPageToken
+			} else {
+				break
+			}
+		}
+
+		if len(activities) == 0 {
+			continue
+		}
+		newLastActivityTime := lastActivityTime
+		// Newest activity is at the end of the list so iterate through the list in reverse.
+		for i := len(activities) - 1; i >= 0; i-- {
```

Comment:
This loop could be dangerous since it just depends how many events are returned from the activities list. If something goes wildly wrong we could spam users with hundreds of notifications. I'm wondering if we find > X number of activities per file that we just batch them into one message or something like that? It's more work again but currently there is no mitigation in place to catch potential issues like that

Reply:
> @BenCookie95 is this something you want to address in this PR or a follow-up?

@wiggin77 I think this is something I'll address in my rate limit PR but I'm leaning towards just adding a max limit to how many activities we fetch. I don't want to add a mechanism to come back later because it does not solve the spamming issues where we subject someone to loads of notifications. Batching multiple comments into one message might require some UX input, not sure if we have a suitable frontend component right now. Can we just send a message to say there has been activity in the document and they should check it?

Reply:
> If there are more than 5 activities can we just send a message to say there has been activity in the document and they should check it?

Sounds reasonable to me.

Reply:
Ok, I'm going to address that in another PR